node/core/approval-voting/src/backend.rs

```diff
@@ -0,0 +1,305 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use kvdb::{DBTransaction, KeyValueDB};
+use polkadot_node_subsystem::{SubsystemResult};
+use polkadot_primitives::v1::{BlockNumber, CandidateHash, Hash};
+use parity_scale_codec::{Encode};
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use super::approval_db;
+use super::approval_db::v1::{
+	blocks_at_height_key, block_entry_key, candidate_entry_key, STORED_BLOCKS_KEY, Config,
+};
+use super::ops::{self, StoredBlockRange};
+use super::persisted_entries::{BlockEntry, CandidateEntry};
+
+#[derive(Debug)]
+pub(super) enum BackendWriteOp {
+	WriteStoredBlockRange(StoredBlockRange),
+	WriteBlocksAtHeight(BlockNumber, Vec<Hash>),
+	WriteBlockEntry(BlockEntry),
+	WriteCandidateEntry(CandidateEntry),
+	DeleteBlocksAtHeight(BlockNumber),
+	DeleteBlockEntry(Hash),
+	DeleteCandidateEntry(CandidateHash),
+}
+
+pub(super) struct DbBackend {
+	inner: Arc<dyn KeyValueDB>,
+	pub(super) config: Config,
```

Comment:
Would prefer this to be a getter `fn config(&self) -> &Config` to protect mutability of the config.